mlir/lib/Dialect/Tosa/Transforms/SinkInputOpsThroughConcat.cpp

```diff
@@ -0,0 +1,425 @@
+    for (auto ops : concatOperands) {
+      auto concatOpReplacement = rewriter.create<tosa::ConcatOp>(
+          concatOp.getLoc(), ops, concatOp.getAxis());
+      concatOps.emplace_back(concatOpReplacement);
+    }
+    // leave it generic here to avoid a bunch of different constructors
+    auto *switchedOp =
+        rewriter.create(concatOp->getLoc(), sample->getName().getIdentifier(),
+                        concatOps, concatOp.getType());
+    for (auto attr : sample->getAttrs()) {
+      switchedOp->setAttr(attr.getName(), attr.getValue());
+    }
+    rewriter.replaceOp(concatOp, switchedOp);
+    return success();
+  }
+
+protected:
+  GenericCheckResult<OpT> doGenericChecks(tosa::ConcatOp concatOp) const {
+    OpT sample = nullptr;
+
+    for (auto val : concatOp->getOperands()) {
+      if (!val.hasOneUse())
+        return GenericCheckResult<OpT>{
+            nullptr, "Operands must just connect to this concat."};
+
+      auto op = val.getDefiningOp<OpT>();
+      if (!op) {
+        std::string err;
+        llvm::raw_string_ostream errStream(err);
+        errStream << "Operand is not a " << OpT::getOperationName();
+        return GenericCheckResult<OpT>{nullptr, err.c_str()};
+      }
+
+      if (!sample)
+        sample = op;
+
+      if (!checkForEquality(sample, op))
+        return GenericCheckResult<OpT>{nullptr,
+                                       "Requires all operands to be the same"};
+    }
+
+    if (sample->getNumOperands() == 0) {
+      return GenericCheckResult<OpT>{
+          nullptr, "Requires all operands to have one or more inputs"};
+    }
+
+    return GenericCheckResult<OpT>{sample, ""};
+  }
+
+  virtual CheckResult doExtraChecks(OpT, tosa::ConcatOp) const {
+    return CheckResult{true, ""};
+  }
+};
+
+template <class OpT>
+struct SinkElementwiseBroadcastableOp : public SinkSpecificOp<OpT> {
+  using SinkSpecificOp<OpT>::SinkSpecificOp;
+
+protected:
+  // check that the broadcast happens on another axis
+  CheckResult doExtraChecks(OpT op, tosa::ConcatOp concat) const override {
+    const char *msg = "Operand broadcasts on same axis then concat.";
+    SmallVector<ArrayRef<int64_t>> shapes;
+    for (auto ty : op->getOperandTypes()) {
+      // lifetime bound to underlying ShapedType object
+      auto tenType = dyn_cast<ShapedType>(ty);
+      if (tenType && tenType.hasStaticShape()) {
+        shapes.emplace_back(tenType.getShape());
+      } else {
+        return CheckResult{
+            false, "Check for broadcast on an unshaped or not static type."};
+      }
+    }
+    assert(shapes.size() == op->getNumOperands() &&
+           "Something went wrong with the above loop.");
+    auto first = shapes[0];
+    for (size_t i = 0; i < first.size(); ++i) {
+      for (size_t j = 1; j < shapes.size(); ++j) {
+        if ((first[i] == 1 || shapes[j][i] == 1) && shapes[j][i] != first[i] &&
+            i == concat.getAxis()) {
+          return CheckResult{false, msg};
+        }
+      }
```

Comment:
If I understand it correctly all the shapes on the concat axis dim need  to be 1, or all need to  be not 1.

You do not need the nested loop here:

```suggestion
const auto axis = concat.getAxis();
const auto &first = shapes[0];
const bool firstIsOne = first[axis] == 1;

const bool mismatch =
  llvm::any_of(shapes, [&](const auto &s) {
    return (firstIsOne ^ (s[axis] == 1)); // xor as  we want to trigger the error if only the first or the other is one
  });

if (mismatch)
  return CheckResult{false, msg};
```

Reply:
Good idea. This should be even simpler by counting the ones:
```
    const auto axis = concat.getAxis();
    // check that the ranks on the axis that the concat uses is not one for all
    // ops or one for all ops
    const auto oneDimensions = llvm::count(
        llvm::map_range(shapes, [&](const auto &s) { return s[axis] == 1; }),
        true);
    if (oneDimensions == shapes.size() || oneDimensions == 0) {
      return CheckResult{false, msg};
    }
```

Reply:
Even simpler by using `count_if` instead of map_range

Reply:
Will do!